### ISSUE TITLE
signal 11 (SIGSEGV) with php-fpm7.3 in debian stretch after Jul 3 upgrade

### ISSUE THREAD

**Issue by dimitris-k** [phpredis/phpredis#1807]
I am using the https://packages.sury.org/php repo.

After the last upgrade to `5.3.0+4.3.0-1+0~20200703.23` from `5.2.1+4.3.0-1+0~20200321.20`
and using `php7.3-fpm 7.3.19-1+0~20200612.60`, the php-fpm processes were killed upon the first request dispatched to them after some seconds, resulting in bad gateway pages...

The local redis server version is `5:5.0.3-3~bpo9+2` listens on a socket and no problems occur when queried using `redis-cli`

some lines from the php7.3-fpm.log:
```
[04-Jul-2020 17:41:00] WARNING: [pool xxx] child 7131 exited on signal 11 (SIGSEGV) after 2.874601 seconds from start
[04-Jul-2020 17:41:00] NOTICE: [pool xxx] child 7137 started
[04-Jul-2020 17:41:01] WARNING: [pool xxx] child 7132 exited on signal 11 (SIGSEGV) after 2.217258 seconds from start
[04-Jul-2020 17:41:01] NOTICE: [pool xxx] child 7138 started
[04-Jul-2020 17:41:01] WARNING: [pool xxx] child 7133 exited on signal 11 (SIGSEGV) after 1.531091 seconds from start
[04-Jul-2020 17:41:01] NOTICE: [pool xxx] child 7139 started
[04-Jul-2020 17:41:01] WARNING: [pool xxx] child 7134 exited on signal 11 (SIGSEGV) after 2.005579 seconds from start
[04-Jul-2020 17:41:01] NOTICE: [pool xxx] child 7140 started
[04-Jul-2020 17:41:02] WARNING: [pool xxx] child 7135 exited on signal 11 (SIGSEGV) after 2.518150 seconds from start
[04-Jul-2020 17:41:02] NOTICE: [pool xxx] child 7141 started
[04-Jul-2020 17:41:03] WARNING: [pool xxx] child 7136 exited on signal 11 (SIGSEGV) after 3.495245 seconds from start
[04-Jul-2020 17:41:03] NOTICE: [pool xxx] child 7142 started
[04-Jul-2020 17:41:04] WARNING: [pool xxx] child 7137 exited on signal 11 (SIGSEGV) after 4.009360 seconds from start
```
Many entries in the `/var/log/messages` are also there and of the form:
```
Jul  4 17:46:59 vmxx kernel: [ 2677.131804] php-fpm7.3[7818]: segfault at 18 ip 00007feb77915676 sp 00007fff1e4e2838 error 4 in libc-2.24.so[7feb77895000+195000]
Jul  4 17:46:59 vmxx kernel: [ 2677.325693] php-fpm7.3[7819]: segfault at 18 ip 00007feb77915676 sp 00007fff1e4e2838 error 4 in libc-2.24.so[7feb77895000+195000]
Jul  4 17:47:00 vmxx kernel: [ 2678.136570] php-fpm7.3[7820]: segfault at 18 ip 00007feb77915676 sp 00007fff1e4e2838 error 4 in libc-2.24.so[7feb77895000+195000]
Jul  4 17:47:01 vmxx kernel: [ 2679.138752] php-fpm7.3[7821]: segfault at 18 ip 00007feb77915676 sp 00007fff1e4e2838 error 4 in libc-2.24.so[7feb77895000+195000]
``` 

Had to revert back to `files` to keep the sessions and make the hosted sites operational, but this could not be the case for some very large installations I am running.

Also, the previous version has been disappeared from the sury.org repo and could not downgrade...

**Comment by michael-grunder** [comment]
Is this happening in a session handler?

https://github.com/phpredis/phpredis/issues/1805

If not please let me know so I can make sure we fix the issue in 5.3.1.

**Comment by dimitris-k** [comment]
It is happening in a session handler. I just followed the workaround to remove `persistent=1` from the `session.save_path`, but it did not fix the problem...
```
session.save_handler = redis
session.save_path = "unix:///var/run/redis/redis.sock?&weight=1&database=0"
```

**Comment by yatsukhnenko** [comment]
@dimitris-k the problem is in dereferencing null pointer. Could you test changes from 5.3.1 branch and confirm that problem is solved and we may publish release?

**Comment by dimitris-k** [comment]
Sorry @yatsukhnenko , I can't do that right now especially on a production server. I will try tomorrow with our developers...

Thank you for your prompt response!

**Comment by nemphys** [comment]
FYI, getting exactly the same issue with php-fpm 7.0 from the same ppa.

**Comment by linaori** [comment]
I've encountered this issue in my local environment. It seems to be a session issue.
```
PHP 7.3.19-1+ubuntu18.04.1+deb.sury.org+1 (cli) (built: Jun 12 2020 07:48:30) ( NTS )

php-redis/bionic,now 5.3.0+4.3.0-1+ubuntu18.04.1+deb.sury.org+1 amd64
```

### Reproducer
Make a php file and call `session_start()` in here. I have the following config in my `redis.ini`
```
extension=redis.so
session.save_handler = redis
session.save_path = "tcp://127.0.0.1:6379?database=2&weight=1&persistent=1"
```
Then call the php file from the CLI.

Commenting these 2 config lines will no longer trigger a segfault.

### Output
```
$ bin/console # session_start() first line and then causes:
Segmentation fault (core dumped)
```

**Comment by yatsukhnenko** [comment]
@linaori could you test `5.3.1` branch?

**Comment by linaori** [comment]
@yatsukhnenko do you have instructions or a guide on how to do that? I've not tested something like this before

**Comment by yatsukhnenko** [comment]
@linaori sorry, I missed that Michael didn't create separated branch yet and fix is in `develop` branch
To test it you need to clone repository and compile extension from sources https://github.com/phpredis/phpredis/blob/develop/INSTALL.markdown#installation-from-sources

**Comment by linaori** [comment]
Thanks, I will give it a try

**Comment by linaori** [comment]
@yatsukhnenko I've taken the following steps:
 - cloned the repository
 - followed the guide, used no configure arguments
 - had to manually move `redis.so`
 - ran the same test file

Looks like everything in my scenario is working again. To be sure I've also checked the webserver variant. Before reloading it was broken, after reloading the service it was working (mod_php7). 

As far as I'm concerned the patch fixes the bug segfault, thanks!

**Comment by yatsukhnenko** [comment]
@linaori cool, thanks! :+1:

**Comment by jeroenvermeulen** [comment]
We are experiencing this issue on PHP 7.1 in combination with ShopWare 5.5.6 on Ubuntu 18.04 using [Ondřej Surý](http://deb.sury.org)'s packages

**Comment by michael-grunder** [comment]
I'm going to release 5.3.1 tomorrow morning (Pacific Time, US).  I just want to give @yatsukhnenko a chance to give it a once over (although it should be fine).

Edit:  [branch is here](https://github.com/phpredis/phpredis/tree/release/5.3.1)

**Comment by yatsukhnenko** [comment]
@michael-grunder 
```diff
diff --git a/redis_session.c b/redis_session.c
index 6f19634..6f2c5c5 100644
--- a/redis_session.c
+++ b/redis_session.c
@@ -467,12 +467,13 @@ PS_OPEN_FUNC(redis)
             RedisSock *redis_sock;
             char *addr, *scheme;
             size_t addrlen;
-            int port;
+            int port, addr_free = 0;
 
             scheme = url->scheme ? REDIS_URL_STR(url->scheme) : "tcp";
             if (url->host) {
                 port = url->port;
                 addrlen = spprintf(&addr, 0, "%s://%s", scheme, REDIS_URL_STR(url->host));
+                addr_free = 1;
             } else { /* unix */
                 port = 0;
                 addr = REDIS_URL_STR(url->path);
@@ -487,7 +488,8 @@ PS_OPEN_FUNC(redis)
             redis_sock->prefix = prefix;
             redis_sock_set_auth(redis_sock, user, pass);
 
-            efree(addr);
+            if (addr_free) efree(addr);
+            if (persistent_id) zend_string_release(persistent_id);
             if (user) zend_string_release(user);
             if (pass) zend_string_release(pass);
             php_url_free(url);
```

**Comment by HansVanEijsden** [comment]
I am experiencing this issue on PHP 7.4.8, installed from source. I'm very happy to see I'm not the only one. Thanks for fixing, looking forward to the pecl release. 👍🏻 

Edit: 5.3.1 solves the issue completely here. Thank you!

**Comment by michael-grunder** [comment]
Closing as this was fixed in 5.3.1.
